fdbclient/ReadYourWrites.actor.cpp

```diff
@@ -1469,6 +1469,14 @@ void ReadYourWritesTransaction::writeRangeToNativeTransaction( KeyRangeRef const
 	}
 }
 
+void ReadYourWritesTransactionOptions::reset(Transaction const& tr) {
+		memset(this, 0, sizeof(*this));
+		Database cx = tr.getDatabase();
+		timeoutInSeconds = cx->transactionTimeout;
+		maxRetries = cx->transactionMaxRetries;
+		snapshotRywEnabled = cx->snapshotRywEnabled;
```

Comment:
I think here's where the timebomb needs to be setup. Also at transaction construction. There's some weirdness that can happen if the next delay will result in the timebomb being triggered (if that makes sense).

Reply:
If I understand correctly, the problem here is that we need to setup a timebomb if there is a timeout set on the database. However, if the client is going to set a larger timeout via transaction option specifically for this transaction, we don't want to end up accidentally terminating the transaction early because the database timeout has elapsed and the currently unset transaction timeout has not. Is that the idea?

Reply:
The problem alluded to was that in the original code, the timeout actor was only initialized to anything useful when someone set the timeout (i.e., within `setOption`'s `TRANSACTION_TIMEOUT` case in the switch block). The initial version of this PR only set the transaction's `timeoutInSeconds` field, which didn't actually do anything. Then 148a21a made it so that the transaction initializes the `timeoutActor` as well.

I suppose that race you alluded to is still possible. In particular, if:

1. The database transaction timeout option is set to *x* and
2. The transaction's timeout is set to *y* > *x* and
3. The transaction encounters a retriable error after *z* time with *x* < *z* < *y* then
4. The transaction will be (spuriously?) completed with a timeout exception.

I suppose we could fix this by, um, not resetting the timeout actor on `onError` if the transaction's timeout is greater than the database's timeout.

Reply:
Ah ok. And I assume it also addressed whatever weirdness you referred to?

As for the race, I'm not sure what the right thing to do is. Your suggestion works but is an exception to the rule that transaction options don't carry over. The best way that immediately occurs to me to avoid that is to set the default timeout somewhere like `getReadVersion` with the idea that once you've started doing work you are subject to the default timeout until otherwise changed, but I don't really love that idea either.

Reply:
Oh, the weirdness--I think that's actually fine. I was thinking of the case where the transaction times out during the retry delay. Having actually run that code now, I think it should work.

I took a crack at addressing the race condition with df8826115d89df. The value stored within the transaction's timeout variable is now reset during `Transaction::reset`, but the actor itself is not. Then the actor is started during either the first get read version call, the first call to set the `timeout` option, or `onError` if the first error is encountered prior to the first read. (I'm not sure when this third case would happen in real life, but it happened in testing by just throwing `not_committed`.) I was able to verify that I could set transaction timeouts that were greater than the default database timeout and have them honored.

Reply:
Okay, after talking to @ajbeamon and @etschannen offline, it seemed like the cleaner thing to do would be to make the transaction not reset its timeout if the user specified one explicitly, and then if there were an explicit timeout, this would be carried over after the `onError` call. I think that is actually isomorphic to just not reseting the timeout in `reset` ever (as if it's set to the database default, then reseting it would set it...back to the default). There's one exception to this, which is if the user sets a default timeout in the middle of things running, existing transactions will use the old default until they are done rather than being set to the new default on `onError`. I think that's fine (arguably to be expected), and it simplifies the logic to "never change the timeout on a horse in the middle of a stream".

This is incompatible with previous behavior, though, so it will continue to reset the timeout if the API version is less than 610, and one is disallowed from setting the database option if the API version is not at least 610.